fix: hang on payload exceed 1k

xujihui1985 opened the pull request:
## Reason for This PR

the reason for hang on payload is because we changed epoll to mio
that only support edge trigger mode, we used to rely on level trigger
to make sure we not miss some event, but for edge trigger, we must make
sure we handle the event properly


## Description of Changes

change the way to handle accept and read event

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.

imeoer commented:
Can we explain how we handle the event properly in the commit message?

jiangliu commented:
Could you please also to add sign-off?

jiangliu commented on src/server.rs:
```diff
@@ -110,44 +111,49 @@ impl<T: Read + Write + ScmSocket> ClientConnection<T> {
+                    while let Some(request) = self.connection.pop_parsed_request() {
+                        // Add all valid requests to `parsed_requests`.
+                        parsed_requests.push(request);
+                        break 'out;
```
Seems the `break out` should be moved out one level so we could collect all parsed reqquests.

xujihui1985 replied:
yes, this is been fixed in the second commit

xujihui1985 commented:
> 

ok, I will sqush the commits and sign off it

wllenyj commented on src/connection.rs:
```diff
@@ -552,6 +552,11 @@ impl<T: Read + Write + ScmSocket> HttpConnection<T> {
         self.parsed_requests.pop_front()
     }
 
+    /// Returns if parsed_requests is empty
```
The comment should be: `is not empty?`

xujihui1985 replied:
> Can try to increase BUFFER_SIZE to solve it? In the source code `src/connection.rs`, can see a size limit on the read buf.

no, I don't think increase BUFFER_SIZE can solve it.

xujihui1985 replied:
the comment has been fixed

wllenyj commented:
Can try to increase BUFFER_SIZE to solve it? 
In the source code `src/connection.rs`, can see a size limit on the read buf.

xujihui1985 committed:
fix: hang on payload exceed 1k

the reason for hang on payload is because we changed epoll to mio
that only support edge trigger mode, we used to rely on level trigger
to make sure we not miss some event, but for edge trigger, we must make
sure we handle the event properly

we used to handle epoll with level trigger, the epoll event will be notified
until all data was processed. that's different in edge trigger, in edge trigger
event will be only notified once there is data. If we can not process the data
we will miss the event.

this patch fix this issue by add loop around read, and read until all data was processed.
in the case of payload exceed 1k, currently we only read 1024 bytes once there is read event.
so we missed the data that exceed 1k

Signed-off-by: xujihui1985 <xujihui1985@gmail.com>

xujihui1985 committed:
fix: ambiguous comment

Signed-off-by: xujihui1985 <xujihui1985@gmail.com>

xujihui1985 commented:
> Can we explain how we handle the event properly in the commit message?

yes, I improve the commit message

wllenyj commented on src/server.rs:
```diff
@@ -330,6 +341,19 @@ impl HttpServer {
+    fn poll_events(&mut self, events: &mut mio::Events) -> Result<()> {
+        loop {
+            if let Err(e) = self.poll.poll(events, None) {
+                if e.kind() == ErrorKind::Interrupted {
```
Need to judge WouldBlock here?

xujihui1985 replied:
yes, probably yes, according to document from mio, we should handle WouldBlock error, and should not treat this as an error
> If operation fails with [WouldBlock](https://doc.rust-lang.org/nightly/std/io/error/enum.ErrorKind.html#variant.WouldBlock), then the caller should not treat this as an error, but instead should wait until another readiness event is received.

wllenyj commented on src/server.rs:
```diff
@@ -581,33 +602,32 @@ impl HttpServer {
+                    Ok(())
+                })
+            {
+                if e.kind() == ErrorKind::WouldBlock {
```
Need to judge Interrupted here?

xujihui1985 replied:
yes, I think so, thanks for point out

xujihui1985 committed:
fix: handle wouldblock on poll

Signed-off-by: xujihui1985 <xujihui1985@gmail.com>

xujihui1985 committed:
fix: handle Interruptted error on accept syscall

Signed-off-by: xujihui1985 <xujihui1985@gmail.com>
